xdsl/parser.py

```diff
@@ -1,1335 +1,1942 @@
-
-            op = self.parse_op_with_default_format(UnregisteredOp,
-                                                   result_types)
-            op.attributes["op_name__"] = StringAttr.from_str(op_name)
+    def must_parse_function_type(self) -> FunctionType:
+        """
+        Parses function-type:
+
+        viable function types are:
+            (i32)   -> ()
+            ()      -> (i32, i32)
+            (i32, i32) -> ()
+            ()      -> i32
+        Non-viable types are:
+            i32     -> i32
+            i32     -> ()
+
+        Uses type-or-type-list-parens internally
+        """
+        self.must_parse_characters(
+            "(", "First group of function args must start with a `(`")
+
+        args: list[Attribute] = self.must_parse_list_of(
+            self.try_parse_type, "Expected type here!")
+
+        self.must_parse_characters(
+            ")",
+            "Malformed function type, expected closing brackets of argument types!"
+        )
+
+        self.must_parse_characters("->",
+                                   "Malformed function type, expected `->`!")
+
+        return FunctionType.from_lists(
+            args, self.must_parse_type_or_type_list_parens())
+
+    def must_parse_type_or_type_list_parens(self) -> list[Attribute]:
+        """
+        Parses type-or-type-list-parens, which is used in function-type.
+
+        type-or-type-list-parens ::= type | type-list-parens
+        type-list-parens         ::= `(` `)` | `(` type-list-no-parens `)`
+        type-list-no-parens      ::=  type (`,` type)*
+        """
+        if self.tokenizer.next_token_of_pattern("(") is not None:
+            args: list[Attribute] = self.must_parse_list_of(
+                self.try_parse_type, "Expected type here!")
+            self.must_parse_characters(
+                ")", "Unclosed function type argument list!")
         else:
-            if not is_generic_format:
-                op = op_type.parse(result_types, self)
-            else:
-                op = self.parse_op_with_default_format(op_type, result_types)
+            args = [self.try_parse_type()]
+            if args[0] is None:
+                self.raise_error(
+                    "Function type must either be single type or list of types in"
+                    " parenthesis!")
+        return args
+
+    def try_parse_function_type(self) -> FunctionType | None:
+        if not self.tokenizer.starts_with("("):
+            return None
+        with self.tokenizer.backtracking("function type"):
+            return self.must_parse_function_type()
+
+    def must_parse_region_list(self) -> list[Region]:
+        """
+        Parses a sequence of regions for as long as there is a `{` in the input.
+        """
+        regions = []
+        while not self.tokenizer.is_eof() and self.tokenizer.starts_with("{"):
+            regions.append(self.must_parse_region())
+        return regions
+
+    # COMMON xDSL/MLIR code:
```

Comment:
I don't like having this in the BaseParser as this implies what the subclasses that you want to use are exactly. It is the case that this is common code for xDSL and MLIR, but what if I want to build my own parser?

Reply:
I think this comment is poorly worded, and I removed it. I think that if someone wants to implement their own parser, they choose what fits them best as a base class (MLIRParser, XDSLParser or BaseParser) and work from there. If they can't work with those options they probably have to start from scratch anyway.

I never thought that there would be more parsers built on-top of the two, but I guess it should be possible without too much hassle. I was under the impression, that we instead want to *reduce* the number of parser implementations we have down to 1.

Reply:
I agree with that, but if we build a BaseParser and two SubParsers, that feels like a structure that is supposed to be used to implement more parsers. If we say that we will remove the difference between xdsl and mlir syntax and then merge both with the BaseParser, I am fine with leaving this as is.

Reply:
Yeah, we want to remove the number of IR parsers to one (essentially removing the xDSL syntax).
However, people using xDSL might want to define their own parsers on top of `xdsl-opt`, for instance to parse an actual language.

Reply:
But for this reason, it's probable that we want to keep what was there before, instead of requiering people to reuse the `BaseParser` class, since they will define their own parsers anyway.